♻️  : 데이터 레이어 및 작가 검색 뷰모델 코드 부분 리팩토링

ldhldh07 opened the pull request:
## 개요
- data 레이어 및 작가 검색 화면 뷰모델 리팩토링

## 변경 사항
- 뷰모델에서 context를 주입받아서 사용하는 안티패턴 조정
  - 어플리케이션 컨텍스트 사용 및 분리
- 작가 홈 화면에서 주소 호출 재사용 예정
  - 주소 호출 로직 분리
- data 레이어 리팩토링
  - service, repository, source 역할이 일반적이지 않아서 조정

## 체크리스트
- [ ] 코드가 작동하는지 확인했습니다.
- [ ] 테스트를 작성했습니다.
- [ ] 문서를 업데이트했습니다.

## 이슈 번호
- Resolves #42

ldhldh07 committed:
:recycle: : 유틸 함수 분리
- context viewModel에서 제거 목적

ldhldh07 committed:
:recycle: : locationService 분리

ldhldh07 committed:
:recycle: : 좌표로 주소 호출 로직 분리
- data 레이어 리팩토링도 병행

ldhldh07 committed:
:recycle: : 작가 호출 로직 리팩토링
- 기존 data 레이어 구조 일관성 부족으로 리팩토링

ldhldh07 committed:
:recycle: : context 주입 방식 수정
- 로케이션 서비스에서는 context @ApplicationContext로 주입받아 사용

Going777 commented on app/src/main/kotlin/com/hm/picplz/data/service/LocationService.kt:
```diff
@@ -0,0 +1,92 @@
+                Manifest.permission.ACCESS_COARSE_LOCATION
+            ) != PackageManager.PERMISSION_GRANTED
+        ) return
+
```
권한 체크 로직을 별도의 함수로 분리해도 좋을 것 같습니다

Going777 commented on app/src/main/kotlin/com/hm/picplz/data/service/LocationService.kt:
```diff
@@ -0,0 +1,92 @@
+
+        // GPS 제공자 사용 시도
+        if (locationManager!!.isProviderEnabled(gpsProvider)) {
+            // 위치 업데이트 리스너 등록
+            val locationListener = LocationListener { location ->
+                onLocationReceived(LatLng.from(location.latitude, location.longitude))
+            }
+            locationListeners.add(locationListener)
+
+            locationManager!!.requestLocationUpdates(
+                gpsProvider,
+                1000L,
+                1.0f,
+                locationListener
+            )
```
LocationListener 등록 로직이 중복되는 것 같습니다. 별도 함수로 리팩토링 하면 좋을 것 같습니당

Going777 commented on app/src/main/kotlin/com/hm/picplz/utils/DisplayMetricsUtil.kt:
```diff
@@ -0,0 +1,41 @@
+        val maxOffsetX = (screenWidth - padding * 2) / 2
+
+        return OffsetLimits(
+            maxOffset = maxOffsetX,
+            innerCircleMaxOffset = maxOffsetX * 0.75f,
+            outerCircleMinOffset = maxOffsetX * 0.93f
```
상수 값(0.75, 0.93)은 (companion) object로 빼서 관리하면 유지보수가 용이할 것 같습니다

Going777 commented on app/src/main/kotlin/com/hm/picplz/utils/DisplayMetricsUtil.kt:
```diff
@@ -0,0 +1,41 @@
+    fun calculateScreenDistance(point1: Offset, point2: Offset): Float {
+        val dx = point1.x - point2.x
+        val dy = point1.y - point2.y
+        return sqrt(dx * dx + dy * dy)
```
hypot()이라는 함수가 존재합니다!

Going777 reviewed (approved):
확인했습니당~

ldhldh07 committed:
:recycle: : 위치 권한 체크 유틸 함수 분리
- 기존 분리 하지 않은 이유는 분리되어 명시적으로 권한 체크를 하지 않을 경우 requestLocationUpdates에서 린트 오류가 발생하였기 때문
- 해당 문제를 try-catch블록을 넣어서 SecurityException을 발생시키며 해결

ldhldh07 committed:
:recycle: : location 요청 함수 분리
- gps와 네트워크 별 로케이션 요청하는 로직이 같아서 분리 후 재사용
- 로케이션 미확인시 try-catch 처리 또한 해당 함수에 위임

ldhldh07 committed:
:recycle: : 위치 권한 요청 로직 수정
- 화면에 입장할 때가 아닌 위치 요청할때의 로직에 포함시킨 후 화면에 들어갈 때는 위치를 요청하도록 수정

ldhldh07 committed:
:recycle: : 동심원 비율 정의
- companion object로 정의하여 유지보수성 증대

ldhldh07 committed:
:recycle: : 거리 계산에 hypot() 적용

ldhldh07 committed:
:bug: : 충돌 처리시 생긴 오류 해결
- 레이어 위치 수정
